src/ucp/core/ucp_listener.c

```diff
@@ -172,99 +172,239 @@ static void ucp_listener_conn_request_callback(uct_iface_h tl_iface, void *arg,
+    ucs_free(listener->ifaces.wifaces);
 }
 
-ucs_status_t ucp_listener_create(ucp_worker_h worker,
-                                 const ucp_listener_params_t *params,
-                                 ucp_listener_h *listener_p)
+static ucs_status_t
+ucp_listen_on_cm(ucp_listener_h listener, const ucp_listener_params_t *params)
 {
-    ucp_context_h context   = worker->context;
-    ucp_listener_h listener = NULL;
-    int sockaddr_tls        = 0;
-    char saddr_str[UCS_SOCKADDR_STRING_LEN];
-    ucp_tl_resource_desc_t *resource;
-    uct_iface_params_t iface_params;
-    ucp_worker_iface_t *tmp;
-    ucp_rsc_index_t tl_id;
+    ucp_worker_h          worker  = listener->worker;
+    const ucp_rsc_index_t num_cms = ucp_worker_get_cm_num(worker);
+    struct sockaddr       addr    = *params->sockaddr.addr;
+    uct_listener_h        *uct_listeners;
+    uct_listener_params_t uct_params;
+    uct_listener_attr_t   uct_attr;
+    uint16_t              port, listener_port;
+    ucp_rsc_index_t i;
+    char addr_str[UCS_SOCKADDR_STRING_LEN];
     ucs_status_t status;
-    ucp_tl_md_t *tl_md;
-    uint16_t port;
-    int i;
 
-    if (!(params->field_mask & UCP_LISTENER_PARAM_FIELD_SOCK_ADDR)) {
-        ucs_error("Missing sockaddr for listener");
-        return UCS_ERR_INVALID_PARAM;
-    }
+    ucs_assert_always(num_cms > 0);
 
-    UCP_CHECK_PARAM_NON_NULL(params->sockaddr.addr, status, return status);
+    uct_params.field_mask       = UCT_LISTENER_PARAM_FIELD_CONN_REQUEST_CB |
+                                  UCT_LISTENER_PARAM_FIELD_USER_DATA;
+    uct_params.conn_request_cb  = (void *)0xdeadbeaf; /* TODO: ucp_listener_conn_request_cb; */
+    uct_params.user_data        = listener;
 
-    if (ucs_test_all_flags(params->field_mask,
-                           UCP_LISTENER_PARAM_FIELD_ACCEPT_HANDLER |
-                           UCP_LISTENER_PARAM_FIELD_CONN_HANDLER)) {
-        ucs_error("Only one accept handler should be provided");
-        return UCS_ERR_INVALID_PARAM;
+    listener->cms.num_listeners = 0;
+    uct_listeners               = ucs_calloc(num_cms, sizeof(*uct_listeners),
+                                             "uct_listeners_arr");
+    if (uct_listeners == NULL) {
+        ucs_error("Can't allocate memory for UCT listeners array");
+        return UCS_ERR_NO_MEMORY;
     }
 
-    UCS_ASYNC_BLOCK(&worker->async);
+    listener->cms.listeners = uct_listeners;
 
-    listener = ucs_calloc(1, sizeof(*listener), "ucp_listener");
-    if (listener == NULL) {
-        status = UCS_ERR_NO_MEMORY;
-        goto out;
-    }
+    for (i = 0; i < num_cms; ++i) {
+        status = uct_listener_create(worker->cms[i], &addr,
+                                     params->sockaddr.addrlen, &uct_params,
+                                     &uct_listeners[listener->cms.num_listeners]);
+        if (status != UCS_OK) {
+            ucs_debug("failed to create UCT listener on CM %p with address %s status %s",
```

Comment:
maybe print component name? (e.g rdmacm/sockcm)?

Reply:
I would avoid in UCP dependency on UCT header (uct/base/uct_cm.h) or component-query-logic just for debug print

Reply:
component name is part of UCT API , ucp_context->tl_cmpts->attr->name

Reply:
component - yes, but `uct_cm_h` is an opaque object. IMO, back mapping of `worker->cms` to `context->tl_cmpts` for debug print is overkill.